verification/certifyKey_test.go

```diff
@@ -44,18 +100,260 @@ func TestCertifyKey_SimulationMode(t *testing.T) {
+				t.Fatalf("TCG DICE MultiTcbInfo extension is not marked as CRITICAL")
+			}
+			_, err := asn1.Unmarshal(ext.Value, &multiTcbInfo)
+			if err != nil {
+				// TODO: Fail when expected to pass
+				t.Logf("Failed to unmarshal MultiTcbInfo field: %v", err)
+			}
+			break
+		}
+	}
+}
+
+// Check whether certificate extended key usage is as per spec
+// OID for ExtendedKeyUsage Extension: 2.5.29.37
+// The ExtendedKeyUsage extension SHOULD be marked as critical
+// If IsCA = true, the extension SHOULD contain tcg-dice-kp-eca
+// If IsCA = false, the extension SHOULD contain tcg-dice-kp-attestLoc
+func checkCertifyKeyExtendedKeyUsages(t *testing.T, c *x509.Certificate) {
+	t.Helper()
+
+	extKeyUsage := []asn1.ObjectIdentifier{}
+
+	for _, ext := range c.Extensions {
+		if ext.Id.String() == ExtensionExtKeyUsage { // OID for ExtKeyUsage extension
+			// Extract the OID value from the extension
+			_, err := asn1.Unmarshal(ext.Value, &extKeyUsage)
+			if err != nil {
+				t.Fatalf("Failed to unmarshal the Extended Key Usage extension: %v", err)
+			}
+
+			if !ext.Critical {
+				t.Fatalf("The Extended Key Usage extension IS NOT CRITICAL, MUST BE CRITICAL")
+			} else {
+				t.Logf("The Extended Key Usage extension is marked CRITICAL")
+			}
+			break
+		}
+	}
+
+	if len(extKeyUsage) == 0 {
+		t.Fatalf("The Extended Key Usage extension is empty")
+	}
+
+	// Iterate over the OIDs in the ExtKeyUsage extension
+	isExtendedKeyUsageValid := false
+	if c.IsCA {
+		for _, oid := range extKeyUsage {
+			if oid.String() == ExtensionTcgDiceKpEca {
+				isExtendedKeyUsageValid = true
+				t.Logf("CA Certificate contains tcg-dice-kp-eca extended key usage")
+				break
+			}
+		}
+		if !isExtendedKeyUsageValid {
+			t.Fatalf("CA Certificate does not contain tcg-dice-kp-eca extended key usage")
+		}
+	} else {
+		for _, oid := range extKeyUsage {
+			if oid.String() == ExtensionTcgDiceKpAttestLoc {
+				isExtendedKeyUsageValid = true
+				t.Logf("Certificate contains tcg-dice-kp-attestLoc extended key usage")
+				break
+			}
+		}
+		if !isExtendedKeyUsageValid {
+			t.Fatalf("Certificate does not contain tcg-dice-kp-attestLoc extended key usage")
+		}
+	}
+}
+
+// Check whether certificate is valid, else log details of the issues found.
+func checkCertificateValidity(t *testing.T, c *x509.Certificate) {
+	t.Helper()
+
+	// Check if the certificate is currently valid - this is checked internally in x509.Verify()
+	now := time.Now().UTC()
+	if now.Before(c.NotBefore) {
+		t.Fatalf("Certificate is not valid yet, current time : %s, not before: %s", now, c.NotBefore)
+	} else if now.After(c.NotAfter) {
+		t.Fatalf("Certificate is expired, current time : %s, not after: %s", now, c.NotAfter)
+	}
+}
+
+// The KeyUsage extension MUST contain only DigitalSignature
+// If IsCA, KeyUsage extension MUST contain DigitalSignature and KeyCertSign
+func checkCertifyKeyExtensions(t *testing.T, c *x509.Certificate) {
+	t.Helper()
+
+	//Check for keyusage extension
+	if c.IsCA {
```

Comment:
This code can mostly be shared. The only difference is allowedKeyUsages

Reply:
Have cut down the duplicate code